cip/CIP2017-06-18-multiple-graphs.adoc

```diff
@@ -0,0 +1,476 @@
+
+As a consequence of adopting the new the multiple graphs execution model, the semantics of each clause needs to be (re-)define how the execution of the clause transforms all given input query contexts into an output query context
+This CIP proposes to preserves all existing semantics by defining how to simulate the single graph execution model in the multiple graphs execution model as outlined below.
+
+=== Partial query contexts
+
+A query context may not return a source graph, a target graph, or even tabular data.
+Such a query context is called a *partial query context*.
+The following rules are proposed to provide defaults for executing a query (or a clause) on a partial query context:
+
+If the input query context of a query (or a clause) does not return a source graph, the *provided source graph* of that query is taken to be the default graph.
+
+If the input query context of a query (or a clause) does not return a target graph, the *provided target graph* of that query is taken to be the default graph.
+
+If the input query context of a query (or a clause) does not return tabular data, the *provided tabular data* of that query is taken to be a single record without any fields.
+
+These rules ensures that a follow-up query (or a clause) that is executed on a partial query context is always provided with a source graph for graph matching, a target graph for graph manipulation, and tabular data such that it will iterate over at least one row.
+
+=== Query composition
+
+The multiple graphs execution model provides a natural way for the sequential composition of queries:
+
+A query `Q1` whose output signature is an acceptable (in terms of provided bindings) input signature for another query `Q2` may be composed sequentially with `Q2` into a new query `Q3` that first runs `Q1` on the initial query context, next runs `Q2` on the query context returned by `Q1`, and finally returns the query context returned by `Q2`.
+
+This homogenous query composition is enabled by using an uniform query context that is passed between clauses.
+
+Note: The currently drafted subquery CIP proposes a language addition (e.g. `THEN`) for expressing this kind of query composition directly. In terms of this CIP, `THEN` is simply syntactic sugar for `WITH * GRAPHS *`
+
+=== Query combinators
+
+Query combinators only need to handle tabular query contexts in the single graph execution model.
+
+For the multiple property graphs execution model, it is necessary to define how query combinators combine the query contexts of all child queries into a new result query context (c.f. query composition).
+
+This CIP proposes that `UNION` (and `UNION ALL` respectively) compute the union between both the source and target graphs, as well as the tabular data (i.e. the records) of all child queries to combine their query contexts.
```

Comment:
So this means that `q1 UNION q2` followed by `q3` (in a composition) will let `q3` match from the union of `q1`'s and `q2`'s source graphs? And for updating statements, if the target graph(s) of `q1` and `q2` are different, updates made by `q3` will actually no be reflected in those target graphs?